### ISSUE TITLE
Learning rate schedulers are suppose to be applied at different times in training

### ISSUE THREAD

**Issue by ddkang** [pytorch/pytorch#2850]
This is a strange design choice.

From http://pytorch.org/docs/master/optim.html#how-to-adjust-learning-rate

StepLR:
```
>>> scheduler = StepLR(optimizer, step_size=30, gamma=0.1)
>>> for epoch in range(100):
>>>     **scheduler.step()**
>>>     train(...)
>>>     validate(...)
```

ReduceLROnPlateau:
```
>>> optimizer = torch.optim.SGD(model.parameters(), lr=0.1, momentum=0.9)
>>> scheduler = ReduceLROnPlateau(optimizer, 'min')
>>> for epoch in range(10):
>>>     train(...)
>>>     val_loss = validate(...)
>>>     # Note that step should be called after validate()
>>>     **scheduler.step(val_loss)**
```

**Comment by chenzhekl** [comment]
`StepLR` does not depend on any external parameters except for current step, while `ReduceLROnPlateau` is designed to reduce learning rate when a metric has stopped improving. Thus it should be called after `validate()`.

**Comment by ddkang** [comment]
I understand this. However, changing where the LR scheduler is called based on the type of LR is still strange. The simplest fix is to move .step() after validation in the other examples as well,

**Comment by Jiaming-Liu** [comment]
Hi. I'm the puller of lr_schedule. I fully agree with the origin post. However, 
> The simplest fix is to move .step() after validation in the other examples as well
This solution would make the concept of `epoch` become 'last_epoch', which is also not a reasonable design. I failed to come up with a better design, and I believe that the difference of calling time is reasonable and easy to understand -- one depends on index of epoch and another depends on the result of validation.

**Comment by ddkang** [comment]
Let's say I wanted to write code that takes in an arbitrary scheduler. In that case, I need to special case it like this:
```
>>> optimizer = torch.optim.SGD(model.parameters(), lr=0.1, momentum=0.9)
>>> scheduler = ReduceLROnPlateau(optimizer, 'min')
>>> for epoch in range(10):
>>>     if scheduler is not ReduceLROnPlateau:
>>>         scheduler.step()
>>>     train(...)
>>>     val_loss = validate(...)
>>>     if scheduler is ReduceLROnPlateau:
>>>         scheduler.step(val_loss)
```
which doesn't seem ideal. I would prefer to pass in epoch-1 or have epoch refer to the last finished epoch.

**Comment by apaszke** [comment]
you can always do this:
```
optimizer = torch.optim.SGD(model.parameters(), lr=0.1, momentum=0.9)
scheduler = ReduceLROnPlateau(optimizer, 'min')
val_loss = validate(...)
for epoch in range(10):
    scheduler.step(*((val_loss,) if scheduler is not ReduceLROnPlateau else ()))
    train(...)
    val_loss = validate(...)
```
